Source/Tools/FEXServer/Unwind/Unwind_x86_64.cpp

```diff
@@ -0,0 +1,155 @@
+        fmt::print(stderr, "R8:  {:016x} R9:  {:016x} R10: {:016x} R11 {:016x}\n",
+          _context.gregs[FEXCore::x86_64::FEX_REG_R8], _context.gregs[FEXCore::x86_64::FEX_REG_R9],
+          _context.gregs[FEXCore::x86_64::FEX_REG_R10], _context.gregs[FEXCore::x86_64::FEX_REG_R11]);
+        fmt::print(stderr, "R12: {:016x} R13: {:016x} R14: {:016x} R15 {:016x}\n",
+          _context.gregs[FEXCore::x86_64::FEX_REG_R12], _context.gregs[FEXCore::x86_64::FEX_REG_R13],
+          _context.gregs[FEXCore::x86_64::FEX_REG_R14], _context.gregs[FEXCore::x86_64::FEX_REG_R15]);
+
+        fmt::print(stderr, "\n");
+      }
+
+      FEXCore::x86_64::mcontext_t * GetContext() {
+        return &_context;
+      }
+
+      unw_word_t GetReg(unw_regnum_t regnum) {
+        switch (regnum) {
+#define GETREG(x) case UNW_X86_64_##x: return _context.gregs[FEXCore::x86_64::FEX_REG_##x]; break
+          GETREG(RAX);
+          GETREG(RDX);
+          GETREG(RCX);
+          GETREG(RBX);
+          GETREG(RSI);
+          GETREG(RDI);
+          GETREG(RBP);
+          GETREG(RSP);
+          GETREG(R8);
+          GETREG(R9);
+          GETREG(R10);
+          GETREG(R11);
+          GETREG(R12);
+          GETREG(R13);
+          GETREG(R14);
+          GETREG(R15);
+          GETREG(RIP);
+#undef GETREG
+          default:
+            panic("Unhandled access regnum: %d", regnum);
+        }
+        return 0;
+      }
+      void SetReg(unw_regnum_t regnum, unw_word_t Value) {
+        switch (regnum) {
+#define GETREG(x) case UNW_X86_64_##x: _context.gregs[FEXCore::x86_64::FEX_REG_##x] = Value; break
+          GETREG(RAX);
+          GETREG(RDX);
+          GETREG(RCX);
+          GETREG(RBX);
+          GETREG(RSI);
+          GETREG(RDI);
+          GETREG(RBP);
+          GETREG(RSP);
+          GETREG(R8);
+          GETREG(R9);
+          GETREG(R10);
+          GETREG(R11);
+          GETREG(R12);
+          GETREG(R13);
+          GETREG(R14);
+          GETREG(R15);
+          GETREG(RIP);
+#undef GETREG
+          default:
+            panic("Unhandled access regnum: %d", regnum);
+        }
+      }
+
+    private:
+      siginfo_t siginfo;
+      FEXCore::x86_64::mcontext_t _context;
+
+      unw_cursor_t cursor;
+      unw_addr_space_t addr;
+
+      unw_accessors_t accessors {
+        .find_proc_info         = UnwinderFuncs::find_proc_info,
+        .put_unwind_info        = UnwinderFuncs::put_unwind_info,
+        .get_dyn_info_list_addr = UnwinderFuncs::get_dyn_info_list_addr,
+        .access_mem             = UnwinderFuncs::access_mem,
+        .access_reg             = UnwinderFuncs::access_reg,
+        .access_fpreg           = UnwinderFuncs::access_fpreg,
+        .resume                 = UnwinderFuncs::resume,
+        .get_proc_name          = UnwinderFuncs::get_proc_name,
+      };
+  };
+
+#define ACCESSOR_IMPL
+#define UNWINDER_TYPE Unwinder_x86_64
+#include "Unwind/Unwind_accessors.inl"
```

Comment:
Is this really necessary? Wouldn't `virtual` member functions in `Unwinder` allow for using a single set of accessors?

Reply:
We can't use a single set of accessors because type definitions change sizes between x86 and x86-64.
Easiest thing to see is that `unw_word_t` is 4 or 8 bytes depending on target remote archirtecture.

Reply:
We only compile a single set of these accessors on a given architecture anyway, so the type differences don't really matter. (If they did in the future, we could abstract over them easily since `unw_word_t` still always fits into 8 bytes.)

The reason I propose `virtal` member functions is so we can call e.g. `BacktraceHeader` from an `Unwinder*` pointer while `Unwinder_x86` is not defined yet. Specifically, this is what `Unwinder` would look like:
```cpp
class Unwinder {
    virtual void SetReg(unw_regnum_t regnum, unw_word_t Value);

    virtual void BacktraceHeader() = 0;

    void Backtrace() {
        // ...
    }
};
```

The functions in `Unwind_accessors.inl` can then be defined purely in terms of `Unwinder` functions. We wouldn't need X-macros and the `.inl` file would just be a normal header.